### PR TITLE
test: add fixes in the sat resolver

### PR DESCRIPTION
**Pull request by x-hgg-x**
### What does this PR try to resolve?

This is a follow-up of https://github.com/rust-lang/cargo/pull/14614.

### How should we test and review this PR?

Commit 1 removes duplicate variables in the sat resolver.
Commit 2 removes useless clones in the sat resolver.

r? Eh2406

### CODE REVIEW

File: crates/resolver-tests/src/sat.rs [hunk 1]
```diff
@@ -86,16 +87,17 @@ fn create_dependencies_vars<'a>(
...
+        let dep_feature_vars = var_for_is_dependencies_features_used
             .entry(name)
             .or_default()
-            .insert((kind, platform), dep_feature_var_map);
+            .entry((kind, platform))
+            .or_default();
+
+        for &feature_name in dep.features() {
+            if let Entry::Vacant(entry) = dep_feature_vars.entry(feature_name) {
+                entry.insert(solver.new_var());
```
Comment (Eh2406) [inline crates/resolver-tests/src/sat.rs:98]: Why the change?  Is this just to avoid  allocating  several `new_var`'s? Do we really end up with dependencies that  activate the same feature more than once?

Also would this be clearer as `.entry(feature_name).or_insert_with(|| solver.new_var())`?

Comment (x-hgg-x) [reply crates/resolver-tests/src/sat.rs:98]: > Why the change? Is this just to avoid allocating several `new_var`'s?

Yes
 
> Also would this be clearer as `.entry(feature_name).or_insert_with(|| solver.new_var())`?

I didn't use `or_insert_with` because I don't need the returned `Entry`.

Comment (Eh2406) [reply crates/resolver-tests/src/sat.rs:98]: Okay, that makes sense.

> Do we really end up with dependencies that activate the same feature more than once?

For testing purposes I added a `else {todo!()}` to each of these, and all tests pass. So I don't think our test suite hits this case. What am I missing? Do you have plans that will add test code that hits this case? Are there significant downsides to a few calls to `solver.new_var()`? I'm not objecting, I'm just trying to understand.

Comment (x-hgg-x) [reply crates/resolver-tests/src/sat.rs:98]: The `else {todo!()}` will trigger if we have a dependency with a list of features containing duplicates, or if the feature `"dep/feature"` or `"dep?/feature"` is listed multiple times among all features declared in a summary. I will add a test for this.

Comment (x-hgg-x) [reply crates/resolver-tests/src/sat.rs:98]: > Are there significant downsides to a few calls to solver.new_var()?

Maybe not, but removing duplicates will improve the debugging if we want to manually inspect the clauses.

**Comment by Eh2406** [comment]
No need for a test.

@bors r+
